spring-integration-ip/src/test/java/org/springframework/integration/ip/tcp/connection/SocketSupportTests.java

```diff
@@ -392,11 +414,70 @@ public void testNioClientAndServerSSL() throws Exception {
 	}
 
 	@Test
-	public void testNioClientAndServerSSLDifferentContextsLargeDataWithReply() throws Exception {
+	public void testNioClientAndServerSSLDifferentContexts() throws Exception {
+		testNioClientAndServerSSLDifferentContexts(false);
+		try {
+			testNioClientAndServerSSLDifferentContexts(true);
+			fail("expected Exception");
+		}
+		catch (Exception e) {
+			// NOSONAR
+		}
+	}
+
+	private void testNioClientAndServerSSLDifferentContexts(boolean badClient) throws Exception {
 		System.setProperty("javax.net.debug", "all"); // SSL activity in the console
 		TcpNioServerConnectionFactory server = new TcpNioServerConnectionFactory(0);
 		TcpSSLContextSupport serverSslContextSupport = new DefaultTcpSSLContextSupport("server.ks",
 				"server.truststore.ks", "secret", "secret");
+		DefaultTcpNioSSLConnectionSupport tcpNioConnectionSupport =
+				new DefaultTcpNioSSLConnectionSupport(serverSslContextSupport) {
+
+					@Override
+					protected void postProcessSSLEngine(SSLEngine sslEngine) {
+						sslEngine.setNeedClientAuth(true);
+					}
+
+		};
+		tcpNioConnectionSupport.afterPropertiesSet();
+		server.setTcpNioConnectionSupport(tcpNioConnectionSupport);
+		final List<Message<?>> messages = new ArrayList<Message<?>>();
+		final CountDownLatch latch = new CountDownLatch(1);
+		server.registerListener(message -> {
+			messages.add(message);
+			latch.countDown();
+			return false;
+		});
+		server.start();
+		TestingUtilities.waitListening(server, null);
+
+		TcpNioClientConnectionFactory client = new TcpNioClientConnectionFactory("localhost", server.getPort());
+		TcpSSLContextSupport clientSslContextSupport = new DefaultTcpSSLContextSupport(
+				badClient ? "server.ks" : "client.ks",
+				"client.truststore.ks", "secret", "secret");
+		DefaultTcpNioSSLConnectionSupport clientTcpNioConnectionSupport =
+				new DefaultTcpNioSSLConnectionSupport(clientSslContextSupport);
+		clientTcpNioConnectionSupport.afterPropertiesSet();
+		client.setTcpNioConnectionSupport(clientTcpNioConnectionSupport);
+		client.start();
+
+		TcpConnection connection = client.getConnection();
+		connection.send(new GenericMessage<String>("Hello, world!"));
+		assertTrue(latch.await(10, TimeUnit.SECONDS));
+		assertEquals("Hello, world!", new String((byte[]) messages.get(0).getPayload()));
+
+		client.stop();
+		server.stop();
+	}
+
+	@Test
+	public void testNioClientAndServerSSLDifferentContextsLargeDataWithReply() throws Exception {
+		System.setProperty("javax.net.debug", "all"); // SSL activity in the console
+		TcpNioServerConnectionFactory server = new TcpNioServerConnectionFactory(0);
+		TcpSSLContextSupport serverSslContextSupport = new DefaultTcpSSLContextSupport("server.ks",
+				"server.truststore.ks", "secret", "secret") {
+
+		};
```

Comment:
The `DefaultTcpSSLContextSupport` isn't `abstract`.
Why do we need this empty extension?